WordPress/gutenberg
# Theme.json:  Add spacing size presets

## What?
As the first step in implementing [spacing size presets in theme.json](https://github.com/WordPress/gutenberg/issues/39371) this PR adds the initial theme.json settings. This will be followed up by additional PRs that will allow the selection of these values in the post and site editor.

## Why?
Currently spacing design options in the editor like padding, margin, and gap only allow for the setting of custom values. This means that theme authors are not able to specify a set list of spacing sizes in order to keep design consistent in the same way as font sizes can be limited to a list of preset sizes.

Also, without a standard list of space presets it is not possible for user applied spacing to easily transfer across themes, or for spacing values to be maintained when block content and patterns are pasted between sites.

## How?
Based on the discussion on the [related issue](https://github.com/WordPress/gutenberg/issues/39371) this PR adds: 

- A `customSpacingSize` boolean to allow theme authors to disable the setting of custom margin, padding, gap values. The first phase of this is a single toggle for all these values, but if there is demand for independent toggle of these this can be change later to allow either a boolean, or an object with booleans for each of the spacing types.
- A `spacingScale` object which contains a list of values to auto generate an array of spacing sizes. The core theme.json only specifies a `spacingScalee` but themes can specify either `spacingScale` and/or  a `spacingSizes` array. Values added to a theme's `spacingScale` will overwrite the core values, and any hard coded spacing size in a theme's `spacingSizes` array will be merged with the core generated array. Theme values with the same slugs as core will override the  core values. A theme can completely remove the core scale by setting `spacing.spaceScale.steps` to `0`
- A `spacingSizes` array contains the list of preset sizes can be added by themes. As with `customSpacingSize` this is a single array of values for all spacing types. If needed this can also be expanded later to allow for either a single array, or an object with an array of values for each spacing type.

Based on the [discussions here](https://github.com/WordPress/gutenberg/issues/39371), it seems to make the most sense to go with a numeric value for the spacing preset slugs and to provide a default set of 9 sizes in the core theme. This allows for 4 settings either side of a `medium` setting. The slugs for the scale will be `10,20,30,40,50,60,70,80,90`, with `50` as the `medium` point. Using a numeric scale like this allows for:

- Themes to extend the scale at any point by inserting additional slugs between the core ones, and doing allows the midpoint to stay at `50`
- makes it easier to provide closest matches if content is moved to a theme that supports a smaller range of presets as [demonstrated here](https://codepen.io/mrwweb/full/bGYQKae) - thanks to @mrwweb for this demo. 

## Testing Instructions

- Add a post and view in frontend
- View page source and check the 9 `--wp--preset--spacing` css custom properties are set in the header
- Add a `spacingScale` object to the current themes `theme.json` settings.spaces` with  the `operator` set  `+` and check that custom properties change correctly
- Add the following `spacingSizes` array to the current themes `theme.json`  and make sure this is merged with the values generated by core
```
 [
		{
			"size": "clamp(16px, 1.5vw, 20px)",
			"slug": "45",
			"name": "Flexi Normal"
		},
		{
			"size": "clamp(1.4rem, 2.15vw, 4rem)",
			"slug": "55",
			"name": "Flexi Large"
		}
]
```
- In themes theme.json set spacingScale.steps to `0` and check that all of the core presets are removed
- Run unit tests `npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php` and `npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/packages/style-engine/phpunit/class-wp-style-engine-test.php`
- Add the following block markup in the edit code view and save and view in the frontend:

```

```

- Check in the frontend that the block markup has converted the `var:preset|space|20` values into `var(--wp--preset--spacing--20);` in the block styles

## Screenshots 
The following custom css properties should be set:
<img width="295" alt="Screen Shot 2022-06-27 at 2 54 05 PM" src="https://user-images.githubusercontent.com/3629020/175851375-f73c63ef-f26e-4f49-bd75-525a5d5557bc.png">

This 9 point, non-linear scale based on a perfect 5th multiplier results in follow spaces based on 16px base font, which puts `medium` at 24px, which is the current default for block gap:

<img width="158" alt="Screen Shot 2022-06-27 at 3 55 32 PM" src="https://user-images.githubusercontent.com/3629020/175857032-2679d513-379a-41c7-bd1d-713cdc593db8.png">
